include/volume/volume_cooling_nonspherical_gaussians_crhmc.hpp

```diff
@@ -0,0 +1,612 @@
+
+    CRHMCRandomPointGenerator::apply(problem, p, N, walk_length, randPoints,
+                                    push_back_policy, rng, g, f, params, walk, simdLen, raw_output);
+
+    while (!done) {
+        NT new_a = last_a * std::pow(ratio, k);
+
+        auto fnit = fn.begin();
+        for (auto pit = randPoints.begin(); pit != randPoints.end(); ++pit, fnit++) {
+            *fnit = eval_exp(*pit, inv_covariance_matrix, new_a, last_a);
+        }
+
+        std::pair<NT, NT> mv = get_mean_variance(fn);
+
+        // Compute a_{i+1}
+        if (mv.second / (mv.first * mv.first) >= C || mv.first / last_ratio < 1.0 + tol) {
+            if (k != 1.0) {
+                k = k / 2;
+            }
+            done = true;
+        } else {
+            k = 2 * k;
+        }
+
+        last_ratio = mv.first;
+    }
+
+    // Return the new a value as a scalar
+    return last_a * std::pow(ratio, k);
+}
+
+// Compute the sequence of non spherical gaussians
+template<
+    int simdLen,
+    typename Point,
+    typename Polytope,
+    typename NT,
+    typename MT,
+    typename RandomNumberGenerator
+>
+void compute_annealing_schedule(Polytope Pin_copy,
+                                Polytope P_copy,
+                                NT const& ratio,
+                                NT const& C,
+                                NT const& frac,
+                                unsigned int const& N,
+                                unsigned int const& walk_length,
+                                NT const& error,
+                                std::vector<NT>& a_vals,
+                                MT const& inv_covariance_matrix,
+                                RandomNumberGenerator& rng) 
+{
+    
+    typedef typename NonSphericalGaussianFunctor::FunctionFunctor<Point>    Func;
+    typedef typename NonSphericalGaussianFunctor::GradientFunctor<Point>    Grad;
+    typedef typename NonSphericalGaussianFunctor::HessianFunctor<Point>     Hess;
+    typedef typename NonSphericalGaussianFunctor::parameters<NT, Point>     func_params;
+
+    typedef crhmc_input<MT, Point, Func, Grad, ZeroFunctor<Point>> Input;
+    //typedef crhmc_input<MT, Point, Func, Grad, Hess> Input;
+
+    typedef crhmc_problem<Point, Input> CrhmcProblem;   
+
+    typedef ImplicitMidpointODESolver<Point, NT, CrhmcProblem, Grad, simdLen> Solver;
+
+    typedef typename CRHMCWalk::template Walk
+            <
+                    Point,
+                    CrhmcProblem,
+                    RandomNumberGenerator,
+                    Grad,
+                    Func,
+                    Solver
+            > CRHMCWalkType;
+
+    typedef typename CRHMCWalk::template parameters
+          <
+                  NT,
+                  Grad
+          > crhmc_walk_params;
+
+    typedef CrhmcRandomPointGenerator<CRHMCWalkType> CRHMCRandomPointGenerator;
+
+    // compute the first variance
+    // for this we need P
+    auto ball = P_copy.ComputeInnerBall();
```

Comment:
Does `P_copy` should be normalized here?

Reply:
Compute Inner Ball normalizes the polytope within itself. So, I dont think it is necessary. https://github.com/GeomScale/volesti/blob/d424f3e31817cf33099a903ca42706c16627cb4d/include/convex_bodies/hpolytope.h#L143